src/nvim/message.c

```diff
@@ -279,33 +277,35 @@ void trunc_string(char_u *s, char_u *buf, int room, int buflen)
     /* For DBCS going backwards in a string is slow, but
      * computing the cell width isn't too slow: go forward
      * until the rest fits. */
-    n = vim_strsize(s + i);
+    n = (size_t)vim_strsize(s + i);
     while (len + n > room) {
-      n -= ptr2cells(s + i);
-      i += (*mb_ptr2len)(s + i);
+      n -= (size_t)ptr2cells(s + i);
+      i += (size_t)(*mb_ptr2len)(s + i);
     }
   } else if (enc_utf8) {
     /* For UTF-8 we can go backwards easily. */
-    half = i = (int)STRLEN(s);
+    half = i = STRLEN(s);
     for (;; ) {
-      do
-        half = half - (*mb_head_off)(s, s + half - 1) - 1;
-      while (utf_iscomposing(utf_ptr2char(s + half)) && half > 0);
-      n = ptr2cells(s + half);
+      do {
+        half = half - (size_t)(*mb_head_off)(s, s + half - 1) - 1;
+      } while (utf_iscomposing(utf_ptr2char(s + half)) && half > 0);
+      n = (size_t)ptr2cells(s + half);
       if (len + n > room)
         break;
       len += n;
       i = half;
     }
   } else {
-    for (i = (int)STRLEN(s); len + (n = ptr2cells(s + i - 1)) <= room; --i)
+    for (i = STRLEN(s);
+         len + (n = (size_t)ptr2cells(s + i - 1)) <= room; --i) {
```

Comment:
When `i` reaches `0`, on the terminal iteration `ptr2cells()` will be called with `i = SIZE_MAX`, which will dereference an invalid pointer. 

This may not show up in the tests because this is in the `!enc_utf8` branch.